### PR TITLE
wgengine/router: make Windows firewall configuration async

### PR DESCRIPTION
**Pull request by bradfitz**
Updating the Windows firewall is usually reasonably fast, but
sometimes blocks for, 20 seconds, 4 minutes, etc. Not sure why.

Until we understand that's happening, configure it in the background
without blocking the normal control flow.

Updates #785

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/867)

### CODE REVIEW

File: wgengine/router/router_windows.go [hunk 1]
```diff
@@ -160,3 +107,83 @@ func (r *winRouter) Close() error {
...
+		ft.mu.Unlock()
+
+		var cmd *exec.Cmd
+		t0 := time.Now()
+		if val == "" {
+			ft.logf("deleting Tailscale-In firewall rule")
+			cmd = exec.Command("netsh", "advfirewall", "firewall", "delete", "rule", "name=Tailscale-In", "dir=in")
+		} else {
+			ft.logf("setting Tailscale-In firewall IP to %q", val)
+			cmd = exec.Command("netsh", "advfirewall", "firewall", "add", "rule", "name=Tailscale-In", "dir=in", "action=allow", "localip="+val, "profile=private", "enable=yes")
```
Comment (danderson) [inline wgengine/router/router_windows.go:168]: This correctly replaces a rule that already exists? No need to go through a delete+add?

**Review by danderson — APPROVED** [review]

File: wgengine/router/router_windows.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/router/router_windows.go:168]: Whoops. I totally misremembered how it worked. I thought the `name=` was a primary key, but nope:

```
Y:\>netsh advfirewall firewall add rule name=Foo dir=in action=allow localip=1.2.4.5 enable=yes
Ok.


Y:\>netsh advfirewall firewall add rule name=Foo dir=in action=allow localip=1.2.4.6 enable=yes
Ok.


Y:\>netsh advfirewall firewall show rule name=Foo

Rule Name:                            Foo
----------------------------------------------------------------------
Enabled:                              Yes
Direction:                            In
Profiles:                             Domain,Private,Public
Grouping:
LocalIP:                              1.2.4.6/32
RemoteIP:                             Any
Protocol:                             Any
Edge traversal:                       No
Action:                               Allow

Rule Name:                            Foo
----------------------------------------------------------------------
Enabled:                              Yes
Direction:                            In
Profiles:                             Domain,Private,Public
Grouping:
LocalIP:                              1.2.4.5/32
RemoteIP:                             Any
Protocol:                             Any
Edge traversal:                       No
Action:                               Allow
Ok.


Y:\>netsh advfirewall firewall delete rule name=Foo

Deleted 2 rule(s).
Ok.
```

So this'll need a bit of changing.

Comment (bradfitz) [reply wgengine/router/router_windows.go:168]: But I guess that's good for IPv6 support.

**Comment by alexbrainman** [comment]
I tried building tailscale-ipn.exe with this change (to see if I can run it), but it does not compile.

Alex

File: wgengine/router/router_windows.go (hunk 1 above)
Comment (alexbrainman) [reply wgengine/router/router_windows.go:168]: I don't see you change code as @danderson suggested. I guess it is still comming.

**Review by alexbrainman — APPROVED** [review]

**Comment by bradfitz** [comment]
> I tried building tailscale-ipn.exe with this change (to see if I can run it), but it does not compile.

It definitely compiles:

Screen Shot 2020-10-29 at 7 36 45 AM

**Comment by bradfitz** [comment]
PTAL

File: wgengine/router/router_windows.go [hunk 2]
```diff
@@ -160,3 +107,110 @@ func (r *winRouter) Close() error {
...
+		ft.known = (err == nil)
+	}
+}
+
+func strsEqual(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
```
Comment (alexbrainman) [inline wgengine/router/router_windows.go:211]: What  happens if and b are not ordered?

**Review by danderson — APPROVED** [review]

File: wgengine/router/router_windows.go [hunk 3]
```diff
@@ -160,3 +107,110 @@ func (r *winRouter) Close() error {
...
+			d, _ := ft.runFirewall("delete", "rule", "name=Tailscale-In", "dir=in")
+			ft.logf("cleared Tailscale-In firewall rules in %v", d)
+		}
+		var err error
+		for _, cidr := range val {
+			ft.logf("adding Tailscale-In rule to allow %v ...", cidr)
+			var d time.Duration
+			d, err = ft.runFirewall("add", "rule", "name=Tailscale-In", "dir=in", "action=allow", "localip="+cidr, "profile=private", "enable=yes")
+			if err != nil {
+				ft.logf("error adding Tailscale-In rule to allow %v: %v", cidr)
```
Comment (alexbrainman) [inline wgengine/router/router_windows.go:193]: You don't want to print error returned by ft.runFirewall?

**Review by alexbrainman — APPROVED** [review]

**Comment by alexbrainman** [comment]
> It definitely compiles:

This builds tests. I tried building tailscale-ipn.exe so I could try and run it.

Alex

**Comment by bradfitz** [comment]
> This builds tests. I tried building tailscale-ipn.exe so I could try and run it.

Without details, I don't know what to say other than you must've done it wrong, as I built it fine multiple times. I tested the code (which involved compiling it) before mailing this out.

File: wgengine/router/router_windows.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/router/router_windows.go:211]: It returns false. This is just a cheaper version of reflect.DeepEqual specialized for `[]string`. I'll add a comment.

File: wgengine/router/router_windows.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/router/router_windows.go:193]: Good catch. I'm sad that staticcheck & vet failed to catch this. (@danderson @josharian, ideas why?)

File: wgengine/router/router_windows.go (hunk 2 above)
Comment (alexbrainman) [reply wgengine/router/router_windows.go:211]: Imagine that strsEqual parameters set to these values

a = []string{"1.1.1.1", "2.2.2.2"}

and

b := []string{"2.2.2.2", "1.1.1.1"}

strsEqual will return false. But I think it should return true. Do you disagree?

**Pull request merged**